src/mono/mono/mini/intrinsics.c

```diff
@@ -488,6 +522,124 @@ emit_unsafe_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatu
 		EMIT_NEW_BIALU (cfg, ins, OP_COMPARE, -1, args [0]->dreg, args [1]->dreg);
 		EMIT_NEW_UNALU (cfg, ins, OP_PCEQ, dreg, -1);
 		return ins;
+	} else if (!strcmp (cmethod->name, "BitCast")) {
+		g_assert (ctx);
+		g_assert (ctx->method_inst);
+		g_assert (ctx->method_inst->type_argc == 2);
+		g_assert (fsig->param_count == 1);
+
+		// We explicitly do not handle gsharedvt as it is meant as a slow fallback strategy
+		// instead we fallback to the managed implementation which will do the right things
+
+		MonoType *tfrom = ctx->method_inst->type_argv [0];
+		if (mini_is_gsharedvt_variable_type (tfrom)) {
+			return NULL;
+		}
+
+		MonoType *tto = ctx->method_inst->type_argv [1];
+		if (mini_is_gsharedvt_variable_type (tto)) {
+			return NULL;
+		}
+
+		// The underlying API always throws for reference type inputs, so we
+		// fallback to the managed implementation to let that handling occur
+
+		MonoTypeEnum tfrom_type = tfrom->type;
+		if (MONO_TYPE_IS_REFERENCE (tfrom)) {
+			return NULL;
+		}
+
+		MonoTypeEnum tto_type = tto->type;
+		if (MONO_TYPE_IS_REFERENCE (tto)) {
+			return NULL;
+		}
+
+		MonoClass *tfrom_klass = mono_class_from_mono_type_internal (tfrom);
+		if (mono_class_is_nullable (tfrom_klass)) {
+			return NULL;
+		}
+
+		// We also always throw for Nullable<T> inputs, so fallback to the
+		// managed implementation here as well.
+
+		MonoClass *tto_klass = mono_class_from_mono_type_internal (tto);
+		if (mono_class_is_nullable (tto_klass)) {
+			return NULL;
+		}
+
+		// The same applies for when the type sizes do not match, as this will always throw
+		// and so its not an expected case and we can fallback to the managed implementation
+
+		int tfrom_align, tto_align;
+		if (mono_type_size (tfrom, &tfrom_align) != mono_type_size (tto, &tto_align)) {
+			return NULL;
+		}
+
+		// We have several different move opcodes to handle the data depending on the
+		// source and target types, so detect and optimize the most common ones falling
+		// back to what is effectively `ReadUnaligned<TTo>(ref As<TFrom, byte>(ref source))`
+		// for anything that can't be special cased as potentially zero-cost move.
+
+		guint32 opcode = OP_LDADDR;
+		MonoStackType tto_stack = STACK_OBJ;
+
+		if (tfrom_type == MONO_TYPE_I4) {
+			if (tto_type == MONO_TYPE_R4) {
+				opcode = OP_MOVE_I4_TO_F;
+				tto_stack = STACK_R4;
+			} else if (tto_type == MONO_TYPE_I4) {
+				opcode = OP_MOVE;
+				tto_stack = STACK_I4;
+			}
```

Comment:
There seems to be a failure around this path for small types (byte->byte, or ushort->short, etc).

Is there some Mono specific handling that's needed here @lambdageek?

Reply:
Or maybe do I need to explicitly check for `MONO_TYPE_I1/I2/U1/U2/U4/U8`?

Reply:
I don't know, I'm sorry. Though `mini_type_to_stack_type` seems to suggest that your latest idea is on the right track.  From the failures it looks like maybe you need to set `ins->klass` too - it looks like there's some kind of sign extension happening when zero extension is expected.

Reply:
yea, maybe that's it.  compare to how we translate `dup`:

```c
		case MONO_CEE_DUP: {
			MonoInst *temp, *store;
			sp--;
			ins = *sp;
			klass = ins->klass;

			temp = mono_compile_create_var (cfg, type_from_stack_type (ins), OP_LOCAL);
			EMIT_NEW_TEMPSTORE (cfg, store, temp->inst_c0, ins);

			EMIT_NEW_TEMPLOAD (cfg, ins, temp->inst_c0);
			ins->klass = klass;
			*sp++ = ins;

			EMIT_NEW_TEMPLOAD (cfg, ins, temp->inst_c0);
			ins->klass = klass;
			*sp++ = ins;

			inline_costs += 2;
			break;
		}
```